framework/aster-frame/src/sync/wait.rs

```diff
@@ -76,16 +79,16 @@ impl WaitQueue {
                 (timeout.as_millis() as u64 + ms_per_tick - 1) / ms_per_tick
             };
 
-            add_timeout_list(remaining_ticks, waiter.clone(), |timer_call_back| {
-                let waiter = timer_call_back
-                    .data()
-                    .downcast_ref::<Arc<Waiter>>()
-                    .unwrap();
-                waiter.wake_up();
+            add_timeout_list(remaining_ticks, waker.clone(), |timer_call_back| {
+                let waker = timer_call_back.data().downcast_ref::<Arc<Waker>>().unwrap();
+                waker.wake_up();
             })
         });
 
         loop {
+            // Enqueue the waker before checking `cond()` to avoid races
+            self.enqueue(waker.clone());
```

Comment:
Yes, the original code actually introduces the problem. If someone wakes the waiting thread after `enqueue` and before `wait`, the waiting thread will waiting forever. I have toughout of making `enqueue` and `wait` atomic, but I can't think up a good interface.
Moving `enqueue` before checking `cond()` seems to be a good idea.

Thanks for your work!